backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/query/impl/ElasticsearchSearchScroll.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.search.query.impl;
+
+import java.util.List;
+
+import org.hibernate.search.backend.elasticsearch.orchestration.impl.ElasticsearchParallelWorkOrchestrator;
+import org.hibernate.search.backend.elasticsearch.search.query.ElasticsearchSearchResult;
+import org.hibernate.search.backend.elasticsearch.work.builder.factory.impl.ElasticsearchWorkBuilderFactory;
+import org.hibernate.search.backend.elasticsearch.work.impl.ElasticsearchSearchResultExtractor;
+import org.hibernate.search.backend.elasticsearch.work.impl.NonBulkableWork;
+import org.hibernate.search.engine.search.query.SearchScroll;
+import org.hibernate.search.engine.search.query.SearchScrollResult;
+import org.hibernate.search.engine.search.query.spi.SimpleSearchScrollResult;
+import org.hibernate.search.util.common.impl.Futures;
+
+public class ElasticsearchSearchScroll<H> implements SearchScroll<H> {
+
+	private final ElasticsearchWorkBuilderFactory workFactory;
+	private final ElasticsearchParallelWorkOrchestrator queryOrchestrator;
+	private final ElasticsearchSearchResultExtractor<ElasticsearchLoadableSearchResult<H>> searchResultExtractor;
+	private final String scrollId;
+	private final List<H> firstNextHits;
+
+	// no reason to mark it as volatile, ElasticsearchSearchScroll instance is not supposed to be thread safe
+	private boolean firstNext = true;
+
+	public ElasticsearchSearchScroll(ElasticsearchWorkBuilderFactory workFactory, ElasticsearchParallelWorkOrchestrator queryOrchestrator,
+			ElasticsearchSearchResultExtractor<ElasticsearchLoadableSearchResult<H>> searchResultExtractor, String scrollId, List<H> firstNextHits) {
+		this.workFactory = workFactory;
+		this.queryOrchestrator = queryOrchestrator;
+		this.searchResultExtractor = searchResultExtractor;
+		this.scrollId = scrollId;
+		this.firstNextHits = firstNextHits;
+	}
+
+	@Override
+	public void close() {
+		Futures.unwrappedExceptionJoin( queryOrchestrator.submit( workFactory.clearScroll( scrollId ).build() ) );
+	}
+
+	@Override
+	public SearchScrollResult<H> next() {
+		if ( firstNext ) {
+			firstNext = false;
+			return new SimpleSearchScrollResult<>( !firstNextHits.isEmpty(), firstNextHits );
```

Comment:
You can't use `!firstNextHits.isEmpty()` to determine whether there were hits or not.

The point of `hasHits()` in `ScrollResult` is to serve as a loop condition: if there were hits on this page, then there may be more pages and I'll iterate one more time.

Imagine that a page of results contains *only* entities that were deleted from the database. Then the hits will be empty. Now imagine there are actually more hits after that, which were *not* delete from the database.
With `!firstNextHits.isEmpty()`, `hasHits()` will return `false` on the first page of results, the client loop will terminate, and it will never even request the second page.

What you really need here is to check that the hits *before loading* are non-empty. I think adding a getter to `ElasticsearchLoadableSearchResult` that returns `extractedHits.size()` would do the trick.

Reply:
right, it makes sense